docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,166 @@
+management systems.
+
+## Motivation
+
+We need agreement on the API design direction and the semantics of the API to be able to
+implement and grow the flightctl service.
+
+### Goals
+
+* Define the format and behavior of the Device and Fleet APIs
+
+* Allow the management of individual fleetless devices on a smaller scale,
+  as an integration mechanism for services like AAP and ACM.
+
+* Allow the management of fleets of devices on a larger scale, via gitops or
+  API.
+
+### Non-Goals
+
+
+
+## Proposal
+
+### Device API
+
+The device API is divided into multiple endpoints, some user facing, and other device/facing,
+this design facilitates cleaner RBAC control of the endpoint access.
+
+#### `/device/{name}` User/Admin facing API
+
+This endpoint can be accessed by the administrator or external systems to
+manage individual devices, also internally from the fleet specifications
+for devices.
+
+When managing a device through a fleet, flightctl is responsible
+of maintaining the device spec in sync with the fleet spec; in this case
+write access is only possible for the modification of labels and other metadata
+through this API.
+
+The proposed changes include:
+
+* Reincorporating device.spec as a visible part through this fronted API.
+* Adding a managementMode field to the device.spec, this field can be set to "fleet" or "manual".
+* The device.spec.os.image contains a reference to the image to be used for the device.
+* The device.spec.containers contains a list of container patterns to be watched on the device (as in the fleet spec)
+* The device.spec.systemd contains a list of systemd service patterns to be watches on the device (as in the fleet spec)
+* The device.spec.config contains a list of config items
+* The config items could be inline definition of config files (inline)
+* Some config items reference exteral configurations (k8s-secret, gitref- ...)
+* External references (image or config items) must support floating tags (latest, main, v1.0),
+  some external items are actually non-versioned, so are inherently floating.
+
+##### Expected behaviour regarding external references with floating tags
+
+Changes on external references is monitored and acted on (git branches,
+device images, secrets), the device controller should re-evaluate the device
+spec and apply the changes.
+
+Several mechanisms can be used to track external references:
+* Polling
+* Webhooks
+
+#### `/device/{name}/rendered` (device side)
+This endpoint is used by the device agent to retrieve the rendered device spec, this endpoint is
+read-only, each device can only access it's own device rendered spec based on the device
+certificate identification or an administrator (for debug purposes).
+
+The proposed changes include:
+* adding a version field, which identifies the version of the configuration.
+* rendering the config details in yaml format instead of a json blob (better for human readability while debugging)
```

Comment:
Can you elaborate on what  "config details" means.  In general I would expect any changes to wire format to be based on performance. Perhaps a more complete example for the problem would be useful.

Reply:
I think I understand that you want to change the format of the spec files on disk for the agent.

Reply:
I was talking about the API by continuing to use a normal structured schema instead of a json blob for now.

Is it really that different in terms of bytes to use json or yaml?
```bash
[majopela@centauro examples]$ yq -p yaml -o json fleet.yaml -I 0
{"apiVersion":"v1alpha1","kind":"Fleet","metadata":{"name":"default"},"spec":{"selector":{"matchLabels":{"fleet":"default"}},"template":{"metadata":{"labels":{"fleet":"default"}},"spec":{"os":{"image":"quay.io/redhat/rhde:9.2"},"config":[{"name":"base","configType":"GitConfigProviderSpec","gitRef":{"repository":"defaultRepo","targetRevision":"main","path":"/basic-nginx-demo/configuration"}},{"name":"motd-update","configType":"InlineConfigProviderSpec","inline":{"ignition":{"version":"3.4.0"},"storage":{"files":[{"contents":{"source":"data:,This%20system%20is%20managed%20by%20flightctl.%0A"},"mode":422,"overwrite":true,"path":"/etc/motd"}]}}}],"systemd":{"matchPatterns":["chronyd.service"]}}}}}
[majopela@centauro examples]$ yq -p yaml -o json fleet.yaml -I 0 | wc --bytes
703
[majopela@centauro examples]$ cat fleet.yaml | wc --bytes
940
```

well a 25%, not bad... :)

I would expect that when our APIs stabilize, we should move on to a binary protocol like gRPC for the agent side, reducing the footprint. There is a bigger difference in there.

Anyway, I will remove this point. I think JSON is probably a good compromise between size and versatility. Perhaps, as you say, we should write YAML to disk to make it easier to debug while maintaining healthier over-the-wire communication.

Reply:
>  we should move on to a binary protocol like gRPC for the agent side, reducing the footprint. There is a bigger difference in there.

I would like to understand the priority of moving the transport protocol to HTTP/2. Is this a roadmap item? Should we begin describing it? With gRPC-gateway we can support both gRPC and JSON.

Reply:
Ok, but I'd try to render into yaml for readability , or may be when printing to the user or saving to disk we could re-render it in a tabulated format for debugability

Reply:
It would make sense that the cli had support for printing to yaml. I believe that solves this problem without code change to agent/server.

Reply:
Yes I think that would be a good solution based on the conversation with avishay.

Reply:
I would defer gRPC for later, when our APIs stabilise, as gRPC requires tracking of the API versions with more care.